### PR TITLE
Initial HCL Support

### PR DESCRIPTION
**Pull request by lkishalmi**
Well, I've restarted my HCL/Terraform Support efforts. This is just the base lexer supporting `tfvar` files, though the lexer, I hope, would work correctly on any HCL files. Parsers are in progress.

@matthiasblaesing could you have a glimpse on the licensing please?

The Terraform icon svg is coming from https://github.com/PKief/vscode-material-icon-theme/ repository. I kept the original as `vscode-terraform.svg`. Unfortunately it seems if I keep the name as `terraform.svg` the IDE would use that one over the png and fails as the svg has no size information.

### CODE REVIEW

**Comment by matthiasblaesing** [comment]
For the icon, I would suggest this approach:

https://github.com/matthiasblaesing/netbeans/commit/7d522935670f181fdecb1c93a22c835cb1c8c82f

- License information was put directly into the SVG. This should still be ok, as we are required to provide the license, but that is done through the file itself.
- I renamed the svg to match the png and modified the SVG to make it possible to use it as icon source in NetBeans

**Comment by lkishalmi** [comment]
@matthiasblaesing Thanks for the SVG! It seems to work, so I've removed the `licenseinfo.xml` and the derived png.
I think the `nbbuild/licenses/MIT-vscode-material-icon-theme` is still into the play, is it?

**Comment by matthiasblaesing** [comment]
> @matthiasblaesing Thanks for the SVG! It seems to work, so I've removed the `licenseinfo.xml` and the derived png. I think the `nbbuild/licenses/MIT-vscode-material-icon-theme` is still into the play, is it?

I think we have no instance of icons where we only store the SVG version. Providing a png fallback makes it possible for people to run without batik, to keep it common, IMHO you should keep the png.

**Comment by matthiasblaesing** [comment]
For the commit validation problem on JDK 8, I suggest to drop `javac.target` from `project.properties` and change `javac.source` to 8. The javac task is modified to switch to using the `release` flag instead of `target`/`source`.

**Comment by mbien** [comment]
@lkishalmi I think @matthiasblaesing meant to downgrade the property to 8. You can't remove the properties otherwise NB won't know what language level the module has and everything will be red in the editor.

basically:
`javac.source=1.8` and either remove target or set it to the same value which is fine too.

**Comment by lkishalmi** [comment]
Well, this one is far from being complete. I even question some of my design decisions, though the base HCL lexer was as close to the (specs)[https://github.com/hashicorp/hcl/blob/main/hclsyntax/spec.md]  as I could get. 

I'm not so sure that I'd keep the TerraformLexer, most probably it would be better to have parser adjusted highlight for different HCL flavors, but till I get there it is how it is.

The AST classes are also some very early implementations.

Anyway, the implementation works providing syntax highlight and folding for tfvars and tf files:
image

It also adds code templates for Terraform files.

File: ide/languages.hcl/src/org/netbeans/modules/languages/hcl/ast/HCLBlock.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    public HCLBlock(List<HCLIdentifier> declaration) {
+        this.declaration = Collections.unmodifiableList(declaration);
+        StringBuilder idb = new StringBuilder(100);
+        String delim = "";
+        for (HCLIdentifier hi : this.declaration) {
+            idb.append(delim).append(hi.id());
+            delim = ".";
+        }
+        id = idb.toString();
```
Comment (mbien) [inline ide/languages.hcl/src/org/netbeans/modules/languages/hcl/ast/HCLBlock.java:42]: potentially easier with `StringJoiner`?

**Review by mbien — APPROVED** [review]
looks good to me. Haven't reviewed the grammar files.

File: ide/languages.hcl/src/org/netbeans/modules/languages/hcl/AnstractHCLLexer.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+import static org.netbeans.modules.languages.hcl.HCLTokenId.*;
+import static org.netbeans.modules.languages.hcl.grammar.HCLLexer.*;
+
+
+/**
+ *
+ * @author lkishalmi
+ */
+public abstract class AnstractHCLLexer extends AbstractAntlrLexerBridge<HCLHereDocAdaptor, HCLTokenId> {
```
[the remark this answers is not in the document]
Comment (lkishalmi) [reply ide/languages.hcl/src/org/netbeans/modules/languages/hcl/AnstractHCLLexer.java]: Good eyes! Thank you!

File: ide/languages.hcl/src/org/netbeans/modules/languages/hcl/HCLHereDocAdaptor.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    protected void popHereDocVar() {
+        currentHereDocVar = hereDocStack.isEmpty() ? null : hereDocStack.removeFirst();
+    }
+
+    protected boolean heredocEndAhead(String partialHeredoc) {
+        int n = 1;
+        int c = _input.LA(1);
+        while ( c == 9 || c == 32) { // Skip leading space and tabs
```
[the remark this answers is not in the document]
Comment (lkishalmi) [reply ide/languages.hcl/src/org/netbeans/modules/languages/hcl/HCLHereDocAdaptor.java]: Checked the Go parser. You are right, it does unicode whitespace.

File: ide/languages.hcl/src/org/netbeans/modules/languages/hcl/ast/HCLBlock.java (hunk 1 above)
Comment (lkishalmi) [reply ide/languages.hcl/src/org/netbeans/modules/languages/hcl/ast/HCLBlock.java]: Learnt something new today!
Thanks!

**Pull request merged**